readme.md
# Manual de zPod

zPod es un complemento para previsualizar multimedia de manera rápida.

Me ofrecieron la idea de crear un complemento que de manera fácil uno pudiese explorar rápidamente una carpeta con muchos archivos multimedia sin necesidad de entrar en un reproductor.

Esto es útil para aquellas personas que necesiten ver rápidamente el contenido de un archivo multimedia y actuar sobre él, si no es el que se busca pasar rápidamente al siguiente y poderlo previsualizar.

Este complemento puede reproducir hasta 139 extensiones multimedia.

Es valido para extensiones tanto de audio como de video pero dejando claro que si fuese un video solo se escuchara el audio, este complemento no tiene interface y se maneja todo a través de teclas asignadas por el usuario.

Decir que también es valido por si llevamos un NVDA portable y en un equipo que no es nuestro deseamos reproducir algo.

Este complemento no necesita tener nada instalado en el ordenador ni depende de ninguna librería externa al complemento por lo que es valido para previsualizar un archivo multimedia en un momento dado que no tengamos un reproductor común.

## Nota importante del autor

Este complemento no pretende ser un reproductor multimedia, se puede usar para escuchar un archivo multimedia en sí, pero nunca se convertirá en un reproductor común ni se le agregarán nuevas opciones que encaminen al complemento en convertirse en tal.

Viene a cubrir una necesidad de poder ver rápidamente el contenido de un archivo multimedia y nada más.

## Uso del complemento

Este complemento en su primera ejecución no tiene ninguna tecla asignada por lo que tendremos que ir a Preferencias / Gestos de entrada / zPod y configurar a nuestro gusto las teclas que consideremos que vamos a necesitar.

Una vez asignadas la más importante es Ejecutar zPod con la cual ejecutaremos el complemento.

Cuando el foco este encima de un archivo multimedia y presionemos esta combinación empezara a reproducirse, pudiendo actuar sobre el archivo con el resto de combinaciones de teclas.

En caso que el archivo multimedia no fuese compatible se nos informara con los correspondientes mensajes.

No se puede elegir más de un archivo para ser reproducido, si lo hiciéramos se nos informara con un mensaje.

Por lo tanto para poner un ejemplo si estamos en un directorio con muchos archivos multimedia y no sabemos cual es el que deseamos podemos ir explorando el directorio e ir pulsando el iniciar el complemento hasta encontrar el que deseamos. 

Cuando este sonando un archivo y elijamos otro automáticamente el que suena se parara y empezara el que elijamos.

## Gestos de entrada

Trae 8 combinaciones que son las siguientes:

* Ejecutar zPod: Con esta combinación una vez configurada iniciaremos la reproducción del archivo multimedia que tenga el foco.

* Poner en pausa la reproducción: Con esto pondremos en pausa el archivo multimedia que este sonando.

* Detener zPod: Con esto pararemos la reproducción del archivo multimedia que este sonando.

* Obtener tiempo reproducido / total: Con esta combinación se nos informara del tiempo transcurrido de reproducción del archivo multimedia y el tiempo total del archivo multimedia.

* Bajar volumen y Subir volumen: Con esto bajaremos y subiremos el volumen de la reproducción, cuando lo pulsemos se nos informara en que porcentaje esta. Cada modificación del sonido se guardara en el archivo NVDA.ini y quedara establecido para la siguiente vez. Por defecto la primera vez es del 50% del sonido.

* Retroceder la reproducción y Adelantar la reproducción: Con estas combinaciones disminuiremos o aumentaremos el tiempo en el archivo multimedia que este sonando. Por defecto la primera vez es de retroceder 5 segundos y adelantar 15 segundos.

Esto se puede configurar en Preferencias / Opciones / zPod, en el panel de opciones de NVDA podremos elegir el tiempo que deseamos para cada acción desde 1 segundo hasta 15 minutos.

Estas opciones se guardarán igualmente en el archivo NVDA.ini y quedaran establecidas para la siguiente vez.

## Recomendación muy importante para poder configurar el primer gesto de entrada Ejecutar zPod

Por si acaso, lo digo que en ningún momento hay que hacer Control + c para iniciar el complemento, eso lo hace el complemento de manera interna solo hay que usar la combinación que asignemos en el gesto de entrada llamado "Ejecutar zPod".

No emplee en la combinación de teclas para iniciar el complemento ni la tecla control ni símbolos, ya que dará errores por que se mezcla el gesto físico con el virtual que mandamos ósea si tenemos por ejemplo un control + alt +i para iniciar realmente luego se convierte en dos control ya que se entrelaza con el que el complemento de manera interna hace, es decir cuando estamos encima de un archivo multimedia lo que hace es que simula una combinación de teclas al archivo exactamente Ctrl +C, exacto esto copia el archivo al portapapeles y de hay se obtiene la ruta del archivo para luego ser ejecutado por zPot.

En las otras combinaciones de teclas da igual pero en la de iniciar el complemento no.

No te olvides que este complemento ya esta hecho y por lo tanto no va a tener más desarrollo que el necesario para su mantenimiento y que funcione con futuras versiones de NVDA.

## Extensiones soportadas

ac3, a52, eac3, mlp, dts, dts-hd, dtshd, true-hd, thd, truehd, thd+ac3, tta, pcm, wav, aiff, aif, aifc, amr, awb, au, snd, lpcm, yuv, y4m, ape, wv, shn, m2ts, m2t, mts, mtv, ts, tsv, tsa, tts, trp, adts, adt, mpa, m1a, m2a, mp1, mp2, mp3, mpeg, mpg, mpe, mpeg2, m1v, m2v, mp2v, mpv, mpv2, mod, tod, vob, vro, evob, evo, mpeg4, m4v, mp4, mp4v, mpg4, m4a, aac, h264, avc, x264, 264, hevc, h265, x265, 265, flac, oga, ogg, opus, spx, ogv, ogm, ogx, mkv, mk3d, mka, webm, weba, avi, vfw, divx, 3iv, xvid, nut, flic, fli, flc, nsv, gxf, mxf, wma, wm, wmv, asf, dvr-ms, dvr, wtv, dv, hdv, flv, f4v, f4a, qt, mov, hdmov, rm, rmvb, ra, ram, 3ga, 3ga2, 3gpp, 3gp, 3gp2, 3g2, ay, gbs, gym, hes, kss, nsf, nsfe, sap, spc, vgm, vgz, m3u, m3u8, pls, cue

## Agradecimientos y traductores

* Reinier Ginnari: Por comentarme su necesidad y darme la idea. También por darme el nombre del complemento y por sus pruebas del complemento y su retroalimentación para llegar a ser lo que es el complemento.
* Francés: Rémy Ruiz
* Portugués: Ángelo Abrantes
* Turco: : umut korkmaz
* Italiano: Alessio Lenzi
* Árabe: Wafiq Taher
* Ucraniano: Heorhii

# Registro de cambios.
## Versión 0.3.1.

* Agregado idioma Ucraniano

* Arreglo en teclados cirílicos

## Versión 0.3.

* Agregada compatibilidad con NVDA 2023.1.

## Versión 0.2.

* Agregada compatibilidad con NVDA 2022.1.

* Agregado idioma Árabe.

## Versión 0.1.5.

* Agregado idioma Francés y  Portugués (Portugal / Brasil).

## Versión 0.1.

* Versión inicial.

changelog.md
* Agregado idioma Ucraniano
* Arreglo en teclados cirílicos

buildVars.py
# -*- coding: UTF-8 -*-

# Build customizations
# Change this file instead of sconstruct or manifest files, whenever possible.


# Since some strings in `addon_info` are translatable,
# we need to include them in the .po files.
# Gettext recognizes only strings given as parameters to the `_` function.
# To avoid initializing translations in this module we simply roll our own "fake" `_` function
# which returns whatever is given to it as an argument.
def _(arg):
	return arg


# Add-on information variables
addon_info = {
	# add-on Name/identifier, internal for NVDA
	"addon_name": _("zPod"),
	# Add-on summary, usually the user visible name of the addon.
	# Translators: Summary for this add-on
	# to be shown on installation and add-on information found in Add-ons Manager.
	"addon_summary": _("zPod para NVDA"),
	# Add-on description
	# Translators: Long description to be shown for this add-on on add-on information from add-ons manager
	"addon_description": _("Previsualiza archivos multimedia rápidamente"),
	# version
	"addon_version": "0.3.1",
	# Author(s)
	"addon_author": u"Héctor J. Benítez Corredera <xebolax@gmail.com>",
	# URL for the add-on documentation support
	"addon_url": "https://github.com/hxebolax/zPod-para-NVDA",
	# Documentation file name
	"addon_docFileName": "readme.html",
	# Minimum NVDA version supported (e.g. "2018.3.0", minor version is optional)
	"addon_minimumNVDAVersion": "2019.3.0",
	# Last NVDA version supported/tested (e.g. "2018.4.0", ideally more recent than minimum version)
	"addon_lastTestedNVDAVersion": "2023.1.0",
	# Add-on update channel (default is None, denoting stable releases,
	# and for development releases, use "dev".)
	# Do not change unless you know what you are doing!
	"addon_updateChannel": None,
}

import os.path

# Define the python files that are the sources of your add-on.
# You can either list every file (using ""/") as a path separator,
# or use glob expressions.
# For example to include all files with a ".py" extension from the "globalPlugins" dir of your add-on
# the list can be written as follows:
# pythonSources = ["addon/globalPlugins/*.py"]
# For more information on SCons Glob expressions please take a look at:
# https://scons.org/doc/production/HTML/scons-user/apd.html
pythonSources = [os.path.join("addon", "globalPlugins", "zPod", "*.py")]

# Files that contain strings for translation. Usually your python sources
i18nSources = ["buildVars.py"] + pythonSources

# Files that will be ignored when building the nvda-addon file
# Paths are relative to the addon directory, not to the root directory of your addon sources.
excludedFiles = []

# Base language for the NVDA add-on
# If the source code, add-on interface and readme file located in the root folder are written in a language other than english, modify this variable as appropriate.
# For example, set baseLanguage to "es" if your add-on is primarily in spanish
baseLanguage = "es"


addon/globalPlugins/zPod/__init__.py
# -*- coding: utf-8 -*-
# Copyright (C) 2021 Héctor J. Benítez Corredera <xebolax@gmail.com>
# This file is covered by the GNU General Public License.

# import the necessary modules (NVDA)
import globalPluginHandler
import addonHandler
import keyboardHandler
import globalVars
import gui
import ui
import config
from scriptHandler import script
from gui.settingsDialogs import NVDASettingsDialog, SettingsPanel
from gui import guiHelper, nvdaControls
import wx
import time
from datetime import timedelta
import ctypes
import os, sys
from .lib import mpv
from .win32 import win32clipboard as cb
from .pubsub import pub

addonHandler.initTranslation()

# Extensiones
dvd_blu = ["ac3", "a52", "eac3", "mlp", "dts", "dts-hd", "dtshd", "true-hd", "thd", "truehd", "thd+ac3", "tta"]
Uncompressed = ["pcm", "wav", "aiff", "aif", "aifc", "amr", "awb", "au", "snd", "lpcm", "yuv", "y4m"]
free_lossless = ["ape", "wv", "shn", ]
mpeg = ["m2ts", "m2t", "mts", "mtv", "ts", "tsv", "tsa", "tts", "trp", "adts", "adt", "mpa", "m1a", "m2a", "mp1", "mp2", "mp3", "mpeg", "mpg", "mpe", "mpeg2", "m1v", "m2v", "mp2v", "mpv", "mpv2", "mod", "tod", "vob", "vro", "evob", "evo", "mpeg4", "m4v", "mp4", "mp4v", "mpg4", "m4a", "aac", "h264", "avc", "x264", "264", "hevc", "h265", "x265", "265", ]
xiph = ["flac", "oga", "ogg", "opus", "spx", "ogv", "ogm", "ogx"]
matroska = ["mkv", "mk3d", "mka", "webm", "weba"]
misc = ["avi", "vfw", "divx", "3iv", "xvid", "nut", "flic", "fli", "flc", "nsv", "gxf", "mxf", ]
windows = ["wma", "wm", "wmv", "asf", "dvr-ms", "dvr", "wtv"]
dv = ["dv", "hdv", ]
flash = ["flv", "f4v", "f4a"]
quicktime = ["qt", "mov", "hdmov"]
real = ["rm", "rmvb", "ra", "ram"]
tres_gpp = ["3ga", "3ga2", "3gpp", "3gp", "3gp2", "3g2"]
video_game = ["ay", "gbs", "gym", "hes", "kss", "nsf", "nsfe", "sap", "spc", "vgm", "vgz"]
playlist = ["m3u", "m3u8", "pls", "cue"]

listaExtensiones = dvd_blu + Uncompressed + free_lossless + mpeg + xiph + matroska + misc + windows + dv + flash + quicktime + real + tres_gpp + video_game + playlist

dictTiempos = {
	0:1,
	1:5,
	2:15,
	3:30,
	4:60,
	5:300,
	6:600,
	7:900,
}
tiempoDescripcion = [
	_("1 segundo"),
	_("5 segundos"),
	_("15 segundos"),
	_("30 segundos"),
	_("1 minuto"),
	_("5 minutos"),
	_("10 minutos"),
	_("15 minutos"),
]

def initConfiguration():
	confspec = {
		"atras": "integer(default=1, min=0, max=7)",
		"adelante": "integer(default=2, min=0, max=7)",
		"volumen": "integer(default=50, min=0, max=100)",
	}
	config.conf.spec['zPod'] = confspec

def getConfig(key):
	value = config.conf["zPod"][key]
	return value

def setConfig(key, value):
	try:
		config.conf.profiles[0]["zPod"][key] = value
	except:
		config.conf["zPod"][key] = value

initConfiguration()
tempAtras = getConfig("atras")
tempAdelante = getConfig("adelante")
volumenGeneral = getConfig("volumen")

player = mpv.MPV(ytdl=True, video=False)

class GlobalPlugin(globalPluginHandler.GlobalPlugin):
	def __init__(self, *args, **kwargs):
		super(GlobalPlugin, self).__init__(*args, **kwargs)

		if globalVars.appArgs.secure: return

		NVDASettingsDialog.categoryClasses.append(zPodPanel)

# Variables
		pub.subscribe(self.onActualizaTiempo, "tiempo")
		pub.subscribe(self.nulo, "terminado")

		self.reproductor = MPVClass()
		self.tiempoTotal = None
		self.tiempo = _("Sin tiempo")

	def terminate(self):
		NVDASettingsDialog.categoryClasses.remove(zPodPanel)

	def nulo(self, event):
		self.tiempo = _("Sin tiempo")

	def onActualizaTiempo(self, event):
		self.tiempo = "{} / {}".format(str(event), self.tiempoTotal)

	def runPlay(self):
		time.sleep(0.7)
		try:
			keyboardHandler.KeyboardInputGesture.fromName("Control+c").send()
		except:
			keyboardHandler.KeyboardInputGesture.fromName("Control+insert").send() # copiar en ruso
		try:
			cb.OpenClipboard()
		except pywintypes.error:
			cb.CloseClipboard()
			cb.OpenClipboard()

		if cb.IsClipboardFormatAvailable(cb.CF_HDROP):
			clipboard_file_path = cb.GetClipboardData(cb.CF_HDROP)
			cb.CloseClipboard()
			numList = len(clipboard_file_path)
			if numList == 0:
				ui.message(_("No hay nada compatible en el foco"))
				return False
			else:
				if numList == 1:
					if os.path.splitext(clipboard_file_path[0])[-1].lower().replace(".", "") in listaExtensiones:
						return clipboard_file_path[0]
					else:
						ui.message(_("El formato del archivo no es compatible"))
						return False
				else:
					ui.message(_("Ha seleccionado %s ficheros. Seleccione solo un fichero.") % numList)
					return False
		else:
			ui.message(_("No hay nada compatible en el foco"))
			return False

	@script(gesture=None, description= _("Ejecutar zPod"), category= _("zPod"))
	def script_zPodRun(self, event):
		p = self.runPlay()
		if ctypes.windll.user32.OpenClipboard(None):
			ctypes.windll.user32.EmptyClipboard()
			ctypes.windll.user32.CloseClipboard()
		if p == False:
			return
		else:
			self.reproductor.volumen(volumenGeneral)
			self.reproductor.play(p)
			self.duracion = self.reproductor.duracion()
			try:
				tiempo1 = int(self.duracion)
				self.tiempoTotal = str(timedelta(seconds=tiempo1))
			except:
				self.tiempoTotal = None

	@script(gesture=None, description= _("Bajar volumen"), category= _("zPod"))
	def script_bajarVol(self, event):
		global volumenGeneral
		if volumenGeneral == 0:
			ui.message(_("No se puede bajar más el volumen ya se encuentra al 0%"))
		else:
			vol1 = volumenGeneral - 1
			volumenGeneral = vol1
			self.reproductor.volumen(volumenGeneral)
			ui.message(_("{}%").format(volumenGeneral))
			setConfig("volumen", volumenGeneral)

	@script(gesture=None, description= _("Subir volumen"), category= _("zPod"))
	def script_subirVol(self, event):
		global volumenGeneral
		if volumenGeneral == 100:
			ui.message(_("No se puede subir más el volumen ya se encuentra al 100%"))
		else:
			vol1 = volumenGeneral + 1
			volumenGeneral = vol1
			self.reproductor.volumen(volumenGeneral)
			ui.message(_("{}%").format(volumenGeneral))
			setConfig("volumen", volumenGeneral)

	@script(gesture=None, description= _("Poner en pausa la reproducción"), category= _("zPod"))
	def script_pausePlay(self, event):
		self.reproductor.pause()

	@script(gesture=None, description= _("Retroceder la reproducción "), category= _("zPod"))
	def script_atrasar(self, event):
		self.reproductor.atrasar(tempAtras)

	@script(gesture=None, description= _("Adelantar la reproducción "), category= _("zPod"))
	def script_adelantar(self, event):
		self.reproductor.adelantar(tempAdelante)

	@script(gesture=None, description= _("Obtener tiempo reproducido / total"), category= _("zPod"))
	def script_tiempo(self, event):
		ui.message(self.tiempo)

	@script(gesture=None, description= _("Detener zPod"), category= _("zPod"))
	def script_terminar(self, event):
		self.reproductor.stop()
		self.tiempo = _("Sin tiempo")

class zPodPanel(SettingsPanel):

	title=_("zPod")

	def makeSettings(self, sizer):
		helper=guiHelper.BoxSizerHelper(self, sizer=sizer)

		self.choiceAtrasar = helper.addLabeledControl(_("Seleccione el tiempo para retroceder la reproducción:"), wx.Choice, choices=tiempoDescripcion)
		self.choiceAtrasar.Bind(wx.EVT_CHOICE, self.onChoiceAtrasar)

		self.choiceAdelantar = helper.addLabeledControl(_("Seleccione el tiempo para adelantar la reproducción:"), wx.Choice, choices=tiempoDescripcion)
		self.choiceAdelantar.Bind(wx.EVT_CHOICE, self.onChoiceAdelantar)

		indexAtras = tempAtras
		self.choiceAtrasar.Selection = indexAtras
		indexAdelante = tempAdelante
		self.choiceAdelantar.Selection = indexAdelante

		self.onChoiceAtrasar(None)
		self.onChoiceAdelantar(None)

	def onSave(self):
		setConfig("atras", self.choiceAtrasar.Selection)
		setConfig("adelante", self.choiceAdelantar.Selection)

	def onChoiceAtrasar(self, event):
		global tempAtras
		tempAtras = self.choiceAtrasar.Selection

	def onChoiceAdelantar(self, event):
		global tempAdelante
		tempAdelante = self.choiceAdelantar.Selection

class MPVClass():
	def __init__(self):

		self.pausa = False
		self.player = player

	def atrasar(self, valor):
		try:
			temp = dictTiempos.get(valor)
			self.player.seek(- temp)
		except:
			pass

	def adelantar(self, valor):
		try:
			self.player.seek(dictTiempos.get(valor))
		except:
			pass

	def volumen(self, valor):
		self.player.volume = valor

	def pause(self):
		if self.pausa:
			self.pausa = False
			self.player.pause = self.pausa
		else:
			self.pausa = True
			self.player.pause = self.pausa

	def stop(self):
		self.player.command('stop')
		self.pausa = False

	def play(self, archivo):
		self.pausa = False
		self.player.play(archivo)

	def duracion(self):
		time.sleep(0.2)
		return self.player.duration

	@player.property_observer('time-pos')
	def observe_time_pos(_name, value):
		if value == None:
			pass
		else:
			f = (lambda x, y: (int(x), int(x*y) % y/y))(float(value), 1e0)
			x=timedelta(seconds=f[0])
			wx.CallAfter(pub.sendMessage, "tiempo", event=x)

	@player.event_callback('end_file')
	def end_file_event(event):
		try:
			wx.CallAfter(pub.sendMessage, "terminado", event=None)
		except:
			pass
